scripts/cluster_buildings.py

```diff
@@ -2,152 +2,144 @@
-    points = []
-    building_ids = []
-    for feature in data["features"]:
-        if feature["geometry"]["type"] == "Point":
-            points.append(feature["geometry"]["coordinates"])
-            building_ids.append(feature["properties"]["id"])
-
-    # Convert points to NumPy array for use with scikit-learn
-    points = np.array(points)
-
-    # Perform k-means clustering
-    kmeans = KMeans(n_clusters=n_clusters, random_state=0).fit(points)
-
-    # Compute centroids of each cluster
+    cleaned_buildings = buildings_classification(input_filepath, crs, house_area_limit)
+    # cleaned_buildings = gpd.read_file(input_filepath)
+    centroids_building = [
+        (row.geometry.centroid.x, row.geometry.centroid.y)
+        for row in cleaned_buildings.itertuples()
+    ]
+    centroids_building = np.array(centroids_building)
+    kmeans = KMeans(n_clusters=n_clusters, random_state=0).fit(centroids_building)
     centroids = kmeans.cluster_centers_
 
-    # Select most central point in each cluster
     central_points = []
     for i in range(kmeans.n_clusters):
-        cluster_points = points[kmeans.labels_ == i]
+        cluster_points = centroids_building[kmeans.labels_ == i]
         distances = np.linalg.norm(cluster_points - centroids[i], axis=1)
         central_point_idx = np.argmin(distances)
         central_points.append(cluster_points[central_point_idx])
 
-    # Create GeoJSON feature for each central point
     features = []
-    for i, central_point in enumerate(central_points):
-        feature = {
-            "type": "Feature",
-            "geometry": {"type": "Point", "coordinates": central_point.tolist()},
-            "properties": {"cluster": i},
-        }
-        features.append(feature)
-
-    # Create GeoJSON object with all central point features
-    geojson = {"type": "FeatureCollection", "features": features}
-
-    # Assign each building to its corresponding cluster based on k-means label
-    for i, label in enumerate(kmeans.labels_):
-        feature = data["features"][i]
-        if feature["geometry"]["type"] == "Point":
-            cluster_id = label
-            if "buildings" not in geojson["features"][cluster_id]["properties"]:
-                geojson["features"][cluster_id]["properties"]["buildings"] = []
-            geojson["features"][cluster_id]["properties"]["buildings"].append(
-                feature["properties"]["id"]
+    for i, row in enumerate(cleaned_buildings.itertuples()):
+        if row.geometry.type == "Polygon":
+            cluster_id = kmeans.labels_[i]
+            features.append(
+                {
+                    "properties": {
+                        "id": row.Index,
+                        "cluster_id": cluster_id,
+                        "tags_building": row.tags_building,
+                        "area_m2": row.area_m2,
+                    },
+                    "geometry": row.geometry,
+                }
             )
```

Comment:
As above this can be revised with something similar to below but adapted:
` microgrid_buildings = microgrid_buildings.loc[microgrid_buildings.geometry.type != "Point"]`
However, I'm unsure we need this part

Reply:
This part was to create a new geodataframe where the buildings would be associated with their own cluster but certainly it can be written better and compacted. What do you think if I made a single function with the two ouputs compacted in this way?


```suggestion
def get_central_points_geojson_with_buildings(
    input_filepath, output_filepath_centroids, n_clusters, crs, house_area_limit,output_filepath_buildings
):
    microgrid_buildings = buildings_classification(input_filepath, crs, house_area_limit)
    centroids_building = [
        (row.geometry.centroid.x, row.geometry.centroid.y)
        for row in microgrid_buildings.itertuples()
    ]
    centroids_building = np.array(centroids_building)
    kmeans = KMeans(n_clusters=n_clusters, random_state=0).fit(centroids_building)
    centroids = kmeans.cluster_centers_
    central_points = []

    for i in range(kmeans.n_clusters):
        cluster_points = centroids_building[kmeans.labels_ == i]
        distances = np.linalg.norm(cluster_points - centroids[i], axis=1)
        central_point_idx = np.argmin(distances)
        central_points.append(cluster_points[central_point_idx])
    central_features = []
    for i, central_point in enumerate(central_points):
        central_features.append(
            {
                "geometry": Point(central_point),
                "cluster": i,
            }
        )
    central_features = gpd.GeoDataFrame(central_features,crs=microgrid_buildings.crs)
    central_features.to_file(output_filepath_centroids)

    clusters = []
    for i, row in enumerate(microgrid_buildings.itertuples()):
        cluster_id = kmeans.labels_[i]
        clusters.append(cluster_id)
    
    microgrid_buildings_gdf = gpd.GeoDataFrame(microgrid_buildings, crs=microgrid_buildings.crs)
    microgrid_buildings_gdf.to_file(output_filepath_buildings)
```